File: app/src/api/clients.ts
import { invoke } from "@tauri-apps/api/core";

export interface Client {
  id: number;
  name: string;
  niu: string;
  phone: string;
  email: string;
  address: string;
  notes: string;
  archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface SaveClientPayload {
  name: string;
  niu: string;
  phone: string;
  email: string;
  address: string;
  notes: string;
}

export async function getAllClients(archived = false): Promise<Client[]> {
  return invoke<Client[]>("get_all_clients", { archived });
}

export async function getClientById(id: number): Promise<Client> {
  return invoke<Client>("get_client_by_id", { id });
}

export async function createClient(
  payload: SaveClientPayload,
): Promise<Client> {
  return invoke<Client>("create_client", { payload });
}

export async function updateClient(
  id: number,
  payload: SaveClientPayload,
): Promise<Client> {
  return invoke<Client>("update_client", { id, payload });
}

export async function archiveClient(id: number): Promise<void> {
  return invoke<void>("archive_client", { id });
}

export async function searchClients(query: string): Promise<Client[]> {
  return invoke<Client[]>("search_clients", { query });
}

export async function getSystemClientId(): Promise<number> {
  return invoke<number>("get_system_client_id");
}

File: app/src-tauri/src/commands/clients.rs
use tauri::State;

use crate::database;
use crate::models::client::{Client, SaveClientPayload};
use crate::AppState;

#[tauri::command]
pub fn get_all_clients(state: State<AppState>, archived: bool) -> Result<Vec<Client>, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::get_all_clients(&conn, archived)
}

#[tauri::command]
pub fn get_client_by_id(state: State<AppState>, id: i64) -> Result<Client, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::get_client_by_id(&conn, id)
}

#[tauri::command]
pub fn create_client(state: State<AppState>, payload: SaveClientPayload) -> Result<Client, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::create_client(&conn, &payload)
}

#[tauri::command]
pub fn update_client(
    state: State<AppState>,
    id: i64,
    payload: SaveClientPayload,
) -> Result<Client, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::update_client(&conn, id, &payload)
}

#[tauri::command]
pub fn archive_client(state: State<AppState>, id: i64) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::archive_client(&conn, id)
}

#[tauri::command]
pub fn search_clients(state: State<AppState>, query: String) -> Result<Vec<Client>, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::search_clients(&conn, &query)
}

#[tauri::command]
pub fn get_system_client_id(state: State<AppState>) -> Result<i64, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::clients::get_system_client_id(&conn)
}
